### ISSUE TITLE
Caret lost its meaning

### ISSUE THREAD

**Issue by akimd** [ruby/strscan#95]
Hi!

Is there any way StringScanner could keep `^` meaningful?  As it is, it denotes the current the beginning of the `rest`, but not whether we're at the start of a line.

In the following example, "bar" is not expected to be stripped.
```
$ cat /tmp/foo.rb
require 'strscan'

def strip(str)
  res = +''
  s = StringScanner.new(str)
  while !s.eos?
    res <<
      if s.scan(/foo/)
        ''
      elsif s.scan(/^bar/)
        ''
      else
        s.scan(/./)
        s.matched
      end
  end
  res
end

p strip('foobarbaz')
```
```
$ ruby /tmp/foo.rb
"baz"
```

If there's no way for StringScanner to "fix" the `^` anchor, the documentation should warn about that.
Cheers!

**Comment by kou** [comment]
How about using `beginning_of_line?`?

```ruby
require 'strscan'

def strip(str)
  res = +''
  s = StringScanner.new(str)
  while !s.eos?
    res <<
      if s.scan(/foo/)
        ''
      elsif s.beginning_of_line? and s.scan(/bar/)
        ''
      else
        s.scan(/./)
        s.matched
      end
  end
  res
end

p strip('foobarbaz')
```

**Comment by akimd** [comment]
I totally missed it... Thanks, it does solve my issue!

However the suggestion remains: there's no way to have `^` mean what it means?
And I think a warning (plus a reference to `beginning_of_line?`) would help.

Thanks a lot for the super fast answer.

**Comment by kou** [comment]
> However the suggestion remains: there's no way to have `^` mean what it means?

Right.

> And I think a warning (plus a reference to `beginning_of_line?`) would help.

I think that there is no way to detect `^` from a `Regexp` object...

**Comment by akimd** [comment]
> I think that there is no way to detect ^ from a Regexp object...

I agree on this.  Not to mention that the anchor can be anywhere in the Regexp, not just at its start.  And maybe in an optional part, or an alternation.

I was rather thinking of a way to tell the RE engine that it is not at a beginning of line so `^` should not match.

I don't know its internals well enough, but that path seems way safer and exact to me rather than trying to massage the regex.

**Comment by akimd** [comment]
> And I think a warning (plus a reference to beginning_of_line?) would help.

To be clear: I was referring to the documentation, not a warning from strscan.
